clippy_lints/src/splitting_strings_at_newlines.rs

```diff
@@ -0,0 +1,89 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for usages of `str.trim().split("\n")` and `str.trim().split("\r\n")`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Hard-coding the line endings makes the code less portable. `str.lines` should be used instead.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".trim().split('\n');
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".lines();
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub SPLITTING_STRINGS_AT_NEWLINES,
+    suspicious,
+    "splitting a trimmed string at hard-coded newlines"
+}
+
+declare_lint_pass!(SplittingStringsAtNewlines => [SPLITTING_STRINGS_AT_NEWLINES]);
+
+// Check if an expr is of a type that implements `Deref` with `Target = str`
+fn derefs_into_str(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
+    matches!(
+        cx.tcx.get_diagnostic_item(sym::Deref),
+        Some(def_id) if (
+            implements_trait(cx, cx.typeck_results().expr_ty(expr), def_id, &[])
+            && matches!(
+                cx.get_associated_type(cx.typeck_results().expr_ty_adjusted(expr), def_id, "Target"),
```

Comment:
The adjusted type here is already a `&str`, so I don't think this projection is necessary. Just `cx.typeck_results().expr_ty_adjusted(expr).peel_refs().is_str()` alone should work.

Reply:
OK, replacing `derefs_into_str` with your suggestion works, and I think I figured out what you're talking about -- but I'm not 100% sure.

> The adjusted type here is already a `&str`

I couldn't really find much documentation on "adjusted types" aside from [`rustc_middle::ty::adjustment`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/adjustment/index.html), and that's rather technical. Am I right that the adjusted type is what you get after things like automatic referencing and dereferencing, as well as method resolution has happened? That is, assuming we have some expression `x` and a call `x.trim()`, and the compiler has resolved that `trim` is `str.trim` (because, with some "adjustments", `x` can be turned into `&str`, e.g. because it implements `Deref<str>`), the adjusted type of `x` in `x.trim()` will be `&str`?

Reply:
Yep that sounds right, the adjusted type is the type after autoderef/autoref and other kinds of implicit coercions. In the case of a method call receiver, it's the type the method is defined on, plus references that were inserted for autoref. So `expr_ty_adjusted(receiver).peel_refs()` should be perfect for this